From 676d0798b74a7b1337894304e0d109f546eca2de Mon Sep 17 00:00:00 2001
From: Florian Leimgruber <fleimgruber@google.com>
Date: Mon, 09 Dec 2024 09:17:53 +0000
Subject: [PATCH] Dangling pointer in UpdateRecentVisitsFromHistoryDBTask

UpdateRecentVisitsFromHistoryDBTask represents a DB task that is
executed on a background thread. It holds a raw_ptr<> to
URLIndexPrivateData, which is accessed in DoneRunOnMainThread() after
the task has finished. The pointer becomes dangling if the
URLIndexPrivateData dies before the task completes.

This CL resolves this by replacing the raw_ptr<> with a
scoped_refptr<>. This works, since URLIndexPrivateData already is a
base::RefCountedThreadSafe<>.

[1] https://source.chromium.org/chromium/chromium/src/+/main:components/omnibox/browser/url_index_private_data.cc;l=350;drc=1651676a30cd7abcd177975f7cd0e37bd945f663

Fixed: 41485242
Change-Id: Ifab8c41614790ff9186cb9b5787cd01e62db2a90
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6054360
Reviewed-by: Jennifer Serrano <jennserrano@google.com>
Commit-Queue: Florian Leimgruber <fleimgruber@google.com>
Cr-Commit-Position: refs/heads/main@{#1393551}
---

--- a/components/omnibox/browser/url_index_private_data.cc
+++ b/components/omnibox/browser/url_index_private_data.cc
@@ -69,7 +69,7 @@
 class UpdateRecentVisitsFromHistoryDBTask : public history::HistoryDBTask {
  public:
   explicit UpdateRecentVisitsFromHistoryDBTask(
-      URLIndexPrivateData* private_data,
+      scoped_refptr<URLIndexPrivateData> private_data,
       history::URLID url_id);
   UpdateRecentVisitsFromHistoryDBTask(
       const UpdateRecentVisitsFromHistoryDBTask&) = delete;
@@ -85,7 +85,7 @@
 
   // The URLIndexPrivateData that gets updated after the historyDB
   // task returns.
-  raw_ptr<URLIndexPrivateData, AcrossTasksDanglingUntriaged> private_data_;
+  scoped_refptr<URLIndexPrivateData> private_data_;
   // The ID of the URL to get visits for and then update.
   history::URLID url_id_;
   // Whether fetching the recent visits for the URL succeeded.
@@ -96,9 +96,11 @@
 };
 
 UpdateRecentVisitsFromHistoryDBTask::UpdateRecentVisitsFromHistoryDBTask(
-    URLIndexPrivateData* private_data,
+    scoped_refptr<URLIndexPrivateData> private_data,
     history::URLID url_id)
-    : private_data_(private_data), url_id_(url_id), succeeded_(false) {}
+    : private_data_(std::move(private_data)),
+      url_id_(url_id),
+      succeeded_(false) {}
 
 bool UpdateRecentVisitsFromHistoryDBTask::RunOnDBThread(
     history::HistoryBackend* backend,
